engine/processReports/src/main/scala/pl/touk/nussknacker/processCounts/influxdb/InfluxCountsReporter.scala

```diff
@@ -3,47 +3,58 @@ package pl.touk.nussknacker.processCounts.influxdb
     case RangeCount(fromDate, toDate) => prepareRangeCounts(processId, fromDate, toDate)
-    case ExecutionCount(pointInTime) => queryInflux(processId, None, pointInTime)
-  }
-
-  private def prepareRangeCounts(processId: String, fromDate: LocalDateTime, toDate: LocalDateTime)(implicit ec: ExecutionContext): Future[String => Option[Long]] = {
-    influxBaseReporter.detectRestarts(processId, fromDate, toDate).flatMap {
-      case Nil => queryInflux(processId, Some(fromDate), toDate)
-      case dates => Future.failed(CannotFetchCountsError(s"Counts unavailable, as process was restarted/deployed on " +
-        s" following dates: ${dates.map(_.format(dateTimeFormatter)).mkString(", ")}"))
+    case ExecutionCount(pointInTime) => influxGenerator.queryBySingleDifference(processId, None, pointInTime, metricsConfig)
+  }).map(ProcessBaseCounts).map(_.getCountForNodeId)
+
+  override def close(): Unit = influxGenerator.close()
+
+  private def prepareRangeCounts(processId: String, fromDate: LocalDateTime, toDate: LocalDateTime)(implicit ec: ExecutionContext): Future[Map[String, Long]] = {
+
+    influxGenerator.detectRestarts(processId, fromDate, toDate, metricsConfig).flatMap { restarts =>
+      (restarts, config.queryMode) match {
+        case (_, QueryMode.OnlySumOfDifferences) =>
+          influxGenerator.queryBySumOfDifferences(processId, fromDate, toDate, metricsConfig)
+        case (Nil, QueryMode.SumOfDifferencesForRestarts) =>
+          influxGenerator.queryBySingleDifference(processId, Some(fromDate), toDate, metricsConfig)
+        case (nonEmpty, QueryMode.SumOfDifferencesForRestarts) =>
+          logger.debug(s"Restarts detected: ${nonEmpty.mkString(",")}, querying with differential")
+          influxGenerator.queryBySumOfDifferences(processId, fromDate, toDate, metricsConfig)
+        case (Nil, QueryMode.OnlySingleDifference) =>
+          influxGenerator.queryBySingleDifference(processId, Some(fromDate), toDate, metricsConfig)
+        case (dates, QueryMode.OnlySingleDifference) =>
+          Future.failed(CannotFetchCountsError.restartsDetected(dates))
+        //should not happen, unfortunately scala won't work...
+        case _ => Future.failed(new IllegalArgumentException(s"Unknown QueryMode ${config.queryMode} for ${restarts.size} restarts"))
+      }
     }
-
-  }
-
-  private def queryInflux(processId: String, fromDate: Option[LocalDateTime], toDate: LocalDateTime)(implicit ec: ExecutionContext): Future[String => Option[Long]] = {
-    influxBaseReporter.fetchBaseProcessCounts(processId, fromDate, toDate).map(pbc => nodeId => pbc.getCountForNodeId(nodeId))
   }
 
-  override def close(): Unit = influxBaseReporter.influxGenerator.close()
-
 }
 
 class InfluxCountsReporterCreator extends CountsReporterCreator {
 
   import net.ceedubs.ficus.Ficus._
   import net.ceedubs.ficus.readers.ArbitraryTypeReader._
-
+  import net.ceedubs.ficus.readers.EnumerationReader._
```

Comment:
Is this needed here or is it a leftover after earlier experiments?

Reply:
It is needed, QueryMode is enumeration